pkg/logs/internal/launchers/file/provider/file_provider.go

```diff
@@ -0,0 +1,281 @@
+type FileProvider struct {
+	filesLimit      int
+	wildcardOrder   wildcardOrdering
+	selectionMode   selectionStrategy
+	shouldLogErrors bool
+}
+
+// NewFileProvider returns a new Provider
+func NewFileProvider(filesLimit int, wildcardOrder wildcardOrdering, selectionMode selectionStrategy) *FileProvider {
+	return &FileProvider{
+		filesLimit:      filesLimit,
+		wildcardOrder:   wildcardOrder,
+		selectionMode:   selectionMode,
+		shouldLogErrors: true,
+	}
+}
+
+type matchCount struct {
+	tracked int
+	total   int
+}
+
+// wildcardFileCounter tracks how many files a wildcard source matches, and how many of those are actually tailed.
+type wildcardFileCounter struct {
+	counts map[*sources.LogSource]matchCount
+}
+
+func newWildcardFileCounter() wildcardFileCounter {
+	return wildcardFileCounter{
+		counts: map[*sources.LogSource]matchCount{},
+	}
+}
+
+func (w *wildcardFileCounter) addTracked(src *sources.LogSource) {
+	matchCnt, _ := w.counts[src]
+	matchCnt.tracked++
+	w.counts[src] = matchCnt
+}
+
+func (w *wildcardFileCounter) setTotal(src *sources.LogSource, total int) {
+	matchCnt, _ := w.counts[src]
+	matchCnt.total = total
+	w.counts[src] = matchCnt
+}
+
+// FilesToTail returns all the Files matching paths in sources,
+// it cannot return more than filesLimit Files.
+// Files are collected according to the fileProvider's sortMode and chooseStrategy
+func (p *FileProvider) FilesToTail(inputSources []*sources.LogSource) []*tailer.File {
+	var filesToTail []*tailer.File
+	shouldLogErrors := p.shouldLogErrors
+	p.shouldLogErrors = false // Let's log errors on first run only
+
+	wildcardFileCounter := newWildcardFileCounter()
+	wildcardSources := make([]*sources.LogSource, 0, len(inputSources))
+
+	consumeFiles := func(files []*tailer.File) {
+		// Add each file one by one
+		for j := 0; j < len(files) && len(filesToTail) < p.filesLimit; j++ {
+			file := files[j]
+			filesToTail = append(filesToTail, file)
+			src := file.Source.UnderlyingSource()
+			if config.ContainsWildcard(src.Config.Path) {
+				wildcardFileCounter.addTracked(src)
+			}
+		}
+
+		if len(filesToTail) >= p.filesLimit {
+			status.AddGlobalWarning(
+				openFilesLimitWarningType,
+				fmt.Sprintf(
+					"The limit on the maximum number of files in use (%d) has been reached. Increase this limit (thanks to the attribute logs_config.open_files_limit in datadog.yaml) or decrease the number of tailed file.",
+					p.filesLimit,
+				),
+			)
+		} else {
+			status.RemoveGlobalWarning(openFilesLimitWarningType)
+		}
+	}
+
+	for i := 0; i < len(inputSources); i++ {
+		source := inputSources[i]
+		isWildcardSource := config.ContainsWildcard(source.Config.Path)
+		if isWildcardSource {
+			// if global, save all wildcards for later
+			if p.selectionMode == GlobalSelection {
+				wildcardSources = append(wildcardSources, source)
+				continue
+			} // else if greedy, collect now source-by-source
```

Comment:
Would it be more readable if we hoist this condition to the top (so we have one selectionMode condition, with both branches executing a for loop inside? Maybe as separate methods too?